aws-android-sdk-s3/src/main/java/com/amazonaws/mobileconnectors/s3/transferutility/UploadPartTask.java

```diff
@@ -53,53 +55,75 @@ public UploadPartTask(UploadTask.UploadPartTaskMetadata uploadPartTaskMetadata,
      */
     @Override
     public Boolean call() throws Exception {
-        try {
-            uploadPartTaskMetadata.state = TransferState.IN_PROGRESS;
-            uploadPartRequest.setGeneralProgressListener(new UploadPartTaskProgressListener(uploadTaskProgressListener));
-            final UploadPartResult putPartResult = s3.uploadPart(uploadPartRequest);
-            uploadPartTaskMetadata.state = TransferState.PART_COMPLETED;
-            dbUtil.updateState(uploadPartRequest.getId(), TransferState.PART_COMPLETED);
-            dbUtil.updateETag(uploadPartRequest.getId(), putPartResult.getETag());
-            return true;
-        } catch (final Exception e) {
-            LOGGER.error("Upload part interrupted: " + e);
-            ProgressEvent resetEvent = new ProgressEvent(0);
-            resetEvent.setEventCode(ProgressEvent.RESET_EVENT_CODE);
-            uploadTaskProgressListener.progressChanged(new ProgressEvent(0));
-
-            // Check if network is not connected, set the state to WAITING_FOR_NETWORK.
+        uploadPartTaskMetadata.state = TransferState.IN_PROGRESS;
+        uploadPartRequest.setGeneralProgressListener(uploadPartTaskProgressListener);
+        int retried = 1;
+        while (true) {
             try {
-                if (TransferNetworkLossHandler.getInstance() != null &&
-                    !TransferNetworkLossHandler.getInstance().isNetworkConnected()) {
-                    LOGGER.info("Thread: [" + Thread.currentThread().getId() + "]: Network wasn't available.");
-                    /*
-                     * Network connection is being interrupted. Moving the TransferState
-                     * to WAITING_FOR_NETWORK till the network availability resumes.
-                     */
-                    uploadPartTaskMetadata.state = TransferState.WAITING_FOR_NETWORK;
-                    dbUtil.updateState(uploadPartRequest.getId(), TransferState.WAITING_FOR_NETWORK);
-                    LOGGER.info("Network Connection Interrupted: " +
-                            "Moving the TransferState to WAITING_FOR_NETWORK");
-                    return false;
+                final UploadPartResult putPartResult = s3.uploadPart(uploadPartRequest);
+                setTaskState(TransferState.PART_COMPLETED);
+                dbUtil.updateETag(uploadPartRequest.getId(), putPartResult.getETag());
+                return true;
+            } catch (AbortedException e) {
+                // If request got aborted, operation was paused or canceled. do not retry.
+                LOGGER.debug("Upload part aborted.");
+                resetProgress();
+                return false;
+            } catch (final Exception e) {
```

Comment:
It looks like this is going to retry for any error.   Ideally, we only want to retry for errors that are retryable (e.g. TimeoutException, any 5xx error), and not for errors that we know will just fail (Unauthorized errors, any 4xx error, etc.)

Can this be updated so that we don't retry for 4xx errors?

Reply:
As discussed offline, any 4xx errors will be caught on the request that initiates the multipart upload.  These are not expected from the requests which upload the parts.